blpapi.ts

```diff
@@ -0,0 +1,728 @@
+
+    request(uri: string,
+            requestName: string,
+            request: any,
+            callback: IRequestCallback): void;
+
+    request(uri: string,
+            requestName: string,
+            request: any,
+            identity: IIdentity,
+            callback: IRequestCallback): void;
+
+    request(uri: string,
+            requestName: string,
+            request: any,
+            label: string,
+            callback: IRequestCallback): void;
+
+    request(uri: string,
+            requestName: string,
+            request: any,
+            identity: IIdentity,
+            label: string,
+            callback: IRequestCallback): void;
+
+    // 'request' function definition
+    request(uri: string,
+            requestName: string,
+            request: any,
+            arg3: Object | string | IRequestCallback,
+            arg4?: string | IRequestCallback,
+            arg5?: IRequestCallback): void
+    {
+        var args = Array.prototype.slice.call(arguments);
+        // Get callback and remove it from array.
+        var callback = args.splice(args.length - 1)[0];
+        // Get optional identity and label arguments (idx 3 and 4 if present).
+        var optionalArgs = args.slice(3);
+
+        var executeRequest = (correlatorId: number): void => {
+            var args = [this.session.request,
+                        uri,
+                        requestName,
+                        request,
+                        correlatorId].concat(optionalArgs);
+            this.invoke.apply(this, args);
+        };
+        this.doRequest(uri, requestName, request, executeRequest, callback);
+    }
+
+    generateToken(cb?: (err: any, value: any) => void): Promise<string>
+    {
+        var correlatorId = this.nextCorrelatorId();
+
+        return new Promise<string>((resolve: (token: string) => void,
+                                    reject: (error: Error) => void): void =>
+        {
+            function callback(err: Error, data?: any): void
+            {
+                if (err) {
+                    reject(err);
+                } else {
+                    if (data.hasOwnProperty('token')) {
+                        resolve(data.token);
+                    } else {
+                        reject(new BlpApiError(data));
+                    }
+                }
+            }
+
+            var executeRequest = (correlatorId: number): void =>
+            {
+                this.listen('TokenGenerationFailure',
+                            correlatorId,
+                            (m: any): void =>
+                            {
+                                this.unlisten('TokenGenerationFailure',
+                                              correlatorId);
+                                reject(new BlpApiError(m.data));
+                            });
+
+                this.invoke(this.session.generateToken, correlatorId);
+            };
+
+            this.doRequest('//blp/apiauth',
```

Comment:
The `requestHandler` won't work well with `TokenGenerationRequest` since the event type is `TOKEN_STATUS`, not `RESPONSE`, so we won't be able to cleanup the listener correctly.